### PR TITLE
storage: fixes replica processing by the replica queue

### LINKED ISSUE #36623: stability: long-period of over-replicated ranges during rebalancing
Thanks to @bdarnell's over-replication metric I noticed that during decommissioning we have a long period of time with over-replicated ranges: 

image

This is troubling due to our lack of atomic rebalancing (https://github.com/cockroachdb/cockroach/issues/12768).

v19.1.0-beta.20190318-701-g4eacf37

*Reproduction*
set -euxo pipefail

export CLUSTER=andy-okr4
roachprod create $CLUSTER -n 10 --gce-machine-type=n1-standard-16 --geo --gce-zones=us-east1-b --gce-zones=us-east1-c --gce-zones=us-east1-d
roachprod run $CLUSTER --  "sudo umount /mnt/data1; sudo mount -o discard,defaults,nobarrier /mnt/data1/; mount | grep /mnt/data1"
roachprod stage $CLUSTER:1,3-10 cockroach
roachprod stage $CLUSTER:2 workload
roachprod start $CLUSTER:1,3-10 
roachprod adminurl --open $CLUSTER:1
roachprod run $CLUSTER:1 -- "./cockroach workload fixtures import tpcc --warehouses=3000 --db=tpcc"
Roachprod ssh $CLUSTER:3
./cockroach quit --decommission --insecure

### PR DESCRIPTION
**Pull request by darinpp**
Fixes #36623

This changes the way the replicas are processed by the replica queue.
Prior to the fix, the replica queue was processing each replica only once
and then was moving to the next replica in the queue. So replicas that
require more than one action (most often caused by the lack of join
consensus in the underlying consensus protocol which results in multiple
steps to get to the desired state) will wait a full
cycle between the first and the second action. With the change, each replica
is processed multiple times (up to three) before moving to the next one in the
queue

Release note: None

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
The approach looks good, I had some cosmetic comments but nothing major. It seems that https://github.com/cockroachdb/cockroach/issues/36623#issuecomment-480939518 might be worth doing even if the ultimate fix here is what you're already doing in this PR. If the priorities are upside down, we should flip them (best in a separate commit). Any particular reason you didn't do that yet?

Have you thought about testing this? I'll cede that it's going to be tricky and that perhaps we can't integration test it well at the moment, especially because the pass/fail condition is so fluid. But I think we should squeeze out some verification from a unit test at least. Is there a test that does a replica movement (i.e. needs more than one change, you can find by inserting a panic whenever `requeue` comes back as true)?

**Comment by tbg** [comment]
PS this definitely needs a release note, please provide one (update both the commit and PR message with it). I'd say "bug fix"

**Review by andreimatei — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @darinpp)

___
*[pkg/storage/replicate_queue.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAObqCt-IF7AadYJf:-LeSAObqCt-IF7AadYJg:bul23b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L253)):*
> ```Go
> 	}
> 
> 	desc, _ := repl.DescAndZone()
> ```

I think `repl.Desc()` is what you want. And then I bet you can use `%s` when you print it below.
___
*[pkg/storage/replicate_queue.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAiDfDzpksbYvwNcq:-LeSAiDfDzpksbYvwNcr:b-snqynu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L274)):*
> ```Go
> 				}
> 				return err
> 			} else if requeue {
> ```

nit: the previous blocks continues or returns; you don't need the else here and so you can unindent the following block.
___
*[pkg/storage/replicate_queue.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSCGDSBPKaUfD9Ks_Q:-LeSCGDSBPKaUfD9Ks_R:byybqsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L276)):*
> ```Go
> 			} else if requeue {
> 				processCount++
> 				if processCount > 3 {
> ```

Any particular reason for having this limit of 3, as opposed to no limit?
___
*[pkg/storage/replicate_queue.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSA7cD0adWa9MpkvJr:-LeSA7cD0adWa9MpkvJs:b-qml2nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L277)):*
> ```Go
> 				processCount++
> 				if processCount > 3 {
> 					log.Infof(ctx, "Exhausted re-processing attempts #%v for %+v", processCount, desc)
> ```

`#%v` for an int? Use `%d`.
___
*[pkg/storage/replicate_queue.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAxRX6E_a7GcQaZNA:-LeSAxRX6E_a7GcQaZNB:bgv8tq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L279)):*
> ```Go
> 					log.Infof(ctx, "Exhausted re-processing attempts #%v for %+v", processCount, desc)
> 					// Enqueue this replica again to see if there are more changes to be made.
> 					rq.MaybeAddAsync(ctx, repl, rq.store.Clock().Now())
> ```

did you figure out what the deal with this `MaybeAddAsync()` is and why it's needed instead of `MaybeAdd()`? And actually, why not just `Add()` (as we know that we want to queue the sucker)
___
*[pkg/storage/replicate_queue.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAZji5Xrm1VHbpG5C:-LeSAZji5Xrm1VHbpG5D:bdskqwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L282)):*
> ```Go
> 				} else {
> 					log.Infof(ctx, "Re-processing #%v for %+v", processCount, desc)
> 					// Try to pro
> ```

try to pro like a pro

File: pkg/storage/replicate_queue.go [hunk 1]
```diff
@@ -250,25 +250,40 @@ func (rq *replicateQueue) process(
...
-				// If ChangeReplicas failed because the preemptive snapshot failed, we
-				// log the error but then return success indicating we should retry the
-				// operation. The most likely causes of the preemptive snapshot failing are
-				// a declined reservation or the remote node being unavailable. In either
-				// case we don't want to wait another scanner cycle before reconsidering
-				// the range.
-				log.Info(ctx, err)
-				continue
+		processCount := 0
+		for {
```
[the remark this answers is not in the document]
Comment (darinpp) [reply pkg/storage/replicate_queue.go]: done

File: pkg/storage/replicate_queue.go [hunk 2]
```diff
@@ -250,25 +250,40 @@ func (rq *replicateQueue) process(
 		MaxRetries:     5,
 	}
 
+	desc, _ := repl.DescAndZone()
```
[the remark this answers is not in the document]
Comment (darinpp) [reply pkg/storage/replicate_queue.go]: removed as it is redundant. The context is enough.

File: pkg/storage/replicate_queue.go [hunk 3]
```diff
@@ -297,7 +312,9 @@ func (rq *replicateQueue) processOneChange(
 	}
 
 	rangeInfo := rangeInfoForRepl(repl, desc)
-	switch action, _ := rq.allocator.ComputeAction(ctx, zone, rangeInfo); action {
+	action, _ := rq.allocator.ComputeAction(ctx, zone, rangeInfo)
+	log.Infof(ctx, "Computed action %+v for %+v", action, desc)
```
[the remark this answers is not in the document]
Comment (darinpp) [reply pkg/storage/replicate_queue.go]: changed it to `VEventf(1, ...)` and removed the lines that log for each branch of the `switch`

File: pkg/storage/replicate_queue.go [hunk 4]
```diff
@@ -250,25 +250,40 @@ func (rq *replicateQueue) process(
...
+				}
+				return err
+			} else if requeue {
+				processCount++
+				if processCount > 3 {
+					log.Infof(ctx, "Exhausted re-processing attempts #%v for %+v", processCount, desc)
+					// Enqueue this replica again to see if there are more changes to be made.
+					rq.MaybeAddAsync(ctx, repl, rq.store.Clock().Now())
+				} else {
+					log.Infof(ctx, "Re-processing #%v for %+v", processCount, desc)
```
[the remark this answers is not in the document]
Comment (darinpp) [reply pkg/storage/replicate_queue.go]: done

File: pkg/storage/replicate_queue.go [hunk 5]
```diff
@@ -250,25 +250,40 @@ func (rq *replicateQueue) process(
...
+					// case we don't want to wait another scanner cycle before reconsidering
+					// the range.
+					log.Info(ctx, err)
+					continue RETRY
+				}
+				return err
+			} else if requeue {
+				processCount++
+				if processCount > 3 {
+					log.Infof(ctx, "Exhausted re-processing attempts #%v for %+v", processCount, desc)
```
[the remark this answers is not in the document]
Comment (darinpp) [reply pkg/storage/replicate_queue.go]: done

**Review by darinpp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/storage/replicate_queue.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAObqCt-IF7AadYJf:-LeTuD5h6OwrLDwHw5sX:bgjs7bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L253)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think `repl.Desc()` is what you want. And then I bet you can use `%s` when you print it below.
</blockquote></details>

I removed it as the context info is duplicate to this. 
___
*[pkg/storage/replicate_queue.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAiDfDzpksbYvwNcq:-LeTuKK30zHEpoTx9Bab:bsauw5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L274)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: the previous blocks continues or returns; you don't need the else here and so you can unindent the following block.
</blockquote></details>

agreed. I fixed it.
___
*[pkg/storage/replicate_queue.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSCGDSBPKaUfD9Ks_Q:-LeTuPCM8cQcJqffba80:b1ig54b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L276)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Any particular reason for having this limit of 3, as opposed to no limit?
</blockquote></details>

I was thinking that a bug can cause the replica to get into infinite loop where we perform set of actions over and over again. With the limit in place, a replica will only be re-processed three times and then the code will give turn to the next replica in the queue. It is also important to note that this is temporary and as soon as we have the join consensus implemented - there won't be a need to do try to keep processing the same replica sequentially. 
___
*[pkg/storage/replicate_queue.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSA7cD0adWa9MpkvJr:-LeTvRUbEmCPhzK-D1mm:b-7rl0kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L277)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`#%v` for an int? Use `%d`.
</blockquote></details>

done
___
*[pkg/storage/replicate_queue.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAxRX6E_a7GcQaZNA:-LeTvVky7kjCbHDcnrQX:b-az7vdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L279)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

did you figure out what the deal with this `MaybeAddAsync()` is and why it's needed instead of `MaybeAdd()`? And actually, why not just `Add()` (as we know that we want to queue the sucker)
</blockquote></details>

Yes, although I'm not sure that I understand exactly why. The issue seems to be call coming from Raft can result in a deadlock acquiring a mutex.
___
*[pkg/storage/replicate_queue.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSAZji5Xrm1VHbpG5C:-LeTvsogEYQzFjXeJZ6H:brpmqc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L282)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

try to pro like a pro
</blockquote></details>

removed.

**Comment by darinpp** [comment]
I did try to add a test and wasn't able to figure out how to do this easily. The suggestion that you have can be done by exporting `process`. Is that what you had in mind? The problem is that the tests are not in `storage` but in `storage_test`. Not sure if this is the preferred setup or was just done like that initially and is hard to change. Seems like we have other examples where the tests are in the same package as the code being tested so you can do a better testing without exporting the methods. 

> The approach looks good, I had some cosmetic comments but nothing major. It seems that [#36623 (comment)](https://github.com/cockroachdb/cockroach/issues/36623#issuecomment-480939518) might be worth doing even if the ultimate fix here is what you're already doing in this PR. If the priorities are upside down, we should flip them (best in a separate commit). Any particular reason you didn't do that yet?
> 
> Have you thought about testing this? I'll cede that it's going to be tricky and that perhaps we can't integration test it well at the moment, especially because the pass/fail condition is so fluid. But I think we should squeeze out some verification from a unit test at least. Is there a test that does a replica movement (i.e. needs more than one change, you can find by inserting a panic whenever `requeue` comes back as true)?

**Review by andreimatei — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @darinpp, and @tbg)

___
*[pkg/storage/replicate_queue.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LeSCGDSBPKaUfD9Ks_Q:-LenLd61A0xtYlh_cwg6:b3dtawu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb665de12df05c0da970ccef69009bb9af0cb71/pkg/storage/replicate_queue.go#L276)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

I was thinking that a bug can cause the replica to get into infinite loop where we perform set of actions over and over again. With the limit in place, a replica will only be re-processed three times and then the code will give turn to the next replica in the queue. It is also important to note that this is temporary and as soon as we have the join consensus implemented - there won't be a need to do try to keep processing the same replica sequentially. 
</blockquote></details>

in my opinion, if you try to be too defensive, you end up with code that doesn't make any sense.
I think a good compromise here, that I was just discussing with @tbg too, is to backport this with this limit of 3, but then do without a limit for 19.2
___
*[pkg/storage/replicate_queue.go, line 286 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37403#-LenJWUZ-eJSfePqP40f:-LenJWU_DGbS52NPwexz:bkg98hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b944bd44acfd96c6379687816437eb436efb7a63/pkg/storage/replicate_queue.go#L286)):*
> ```Go
> 
> 			if i == maxAttempts {
> 				log.VEventf(ctx, 1, "Exhausted re-processing attempts #%d", i)
> ```

nit: our logs lines never start with caps

**Pull request merged**